### DISCUSSION TITLE
Change the default file sanitization method

### DISCUSSION THREAD

**Discussion by squatto** [spatie/laravel-medialibrary#2027] [Ideas]
I wanted to open a discussion somewhat-related to PR #1977 I submitted that you marked as "revisit for next major version". Rather than the sanitization method that I suggested in that PR, I would suggest that this be used:

```php
implode('.', array_map([Str::class, 'slug'], explode('.', $filename)))
```

What this does is create filesystem-friendly filenames that better handle filenames that the current sanitization method doesn't handle very well. Since it's common to use periods to separate filename components (`database.production.sql` for example), this method ensures that they are kept in place. This also has the side effect of preserving the file extension.

I grabbed some filenames from my computer and checked out the differences between the current ("old") method and the proposed ("new") method. It only outputs filenames where the old and new methods produce different results:

```php
$filenames = [
  'Dog, Coffee, Halloween Candy - UV Printed Sign - 15x5.pdf',
  'upsell_3.3.5.zip',
  'Atmosphæra Incognita by Neal Stephenson.pdf',
  'Blacksmith Fork Canyon - cabin plot map.tif',
  'bttv_settings.backup',
  'Cedar Springs plot map.tif',
  'CLPD_Pondera_08142020_FromArchive.txt',
  'contact-form-7-paypal-add-on-pro-2.6.13.zip',
  'How to add M4R ringtone to your iPhone.pdf',
  'Tweeka TV - Episode 12 (The Moskau Edition) - Camden at 52.04 to 52.19.mp4',
];

foreach ($filenames as $filename) {
  $old = strtolower(str_replace(['#', '/', '\\', ' '], '-', $filename));
  $new = implode('.', array_map([Str::class, 'slug'], explode('.', $filename)));
  
  if ($old === $new) {
    continue;
  }
  
  echo "file: $filename\n";
  echo "old:  $old\n";
  echo "new:  $new\n";
  echo "\n";
}
```

Here is the output:

```
file: Dog, Coffee, Halloween Candy - UV Printed Sign - 15x5.pdf
old:  dog,-coffee,-halloween-candy---uv-printed-sign---15x5.pdf
new:  dog-coffee-halloween-candy-uv-printed-sign-15x5.pdf

file: upsell_3.3.5.zip
old:  upsell_3.3.5.zip
new:  upsell-3.3.5.zip

file: Atmosphæra Incognita by Neal Stephenson.pdf
old:  atmosphæra-incognita-by-neal-stephenson.pdf
new:  atmosphaera-incognita-by-neal-stephenson.pdf

file: Blacksmith Fork Canyon - cabin plot map.tif
old:  blacksmith-fork-canyon---cabin-plot-map.tif
new:  blacksmith-fork-canyon-cabin-plot-map.tif

file: bttv_settings.backup
old:  bttv_settings.backup
new:  bttv-settings.backup

file: CLPD_Pondera_08142020_FromArchive.txt
old:  clpd_pondera_08142020_fromarchive.txt
new:  clpd-pondera-08142020-fromarchive.txt

file: Tweeka TV - Episode 12 (The Moskau Edition) - Camden at 52.04 to 52.19.mp4
old:  tweeka-tv---episode-12-(the-moskau-edition)---camden-at-52.04-to-52.19.mp4
new:  tweeka-tv-episode-12-the-moskau-edition-camden-at-52.04-to-52.19.mp4
```

As you can see, the current method produces some very strange filenames:

`dog,-coffee,-halloween-candy---uv-printed-sign---15x5.pdf`
`tweeka-tv---episode-12-(the-moskau-edition)---camden-at-52.04-to-52.19.mp4`

One interesting side effect of using `Str::slug()` is that UTF-8 characters are turned into their ASCII equivalents. You can see this with `Atmosphæra Incognita by Neal Stephenson.pdf` - the `æ` becomes `ae`. This could help with compatibility across different filesystems.

In my own application, the current sanitization method has produced tons of strange filenames, despite our use of a pretty standard file naming scheme:

![image](https://user-images.githubusercontent.com/748444/91621220-96358700-e94f-11ea-8fa8-ad162ee0ec19.png)

Thank you for your consideration!

**Comment by freekmurze** [comment]
Seems like a good idea!
